File: Yona/Yona/Stores/TileStore.swift
//
//  TileStore.swift
//  Yona
//
//  Owns the tile list for the UI. Slice 1: load + render. Create/update/delete
//  arrive in later Phase 2 slices.
//

import Foundation
import Observation

@MainActor
@Observable
final class TileStore {
    private(set) var tiles: LoadState<[Tile]> = .idle

    private let repository: SupabaseRepository

    init(repository: SupabaseRepository) {
        self.repository = repository
    }

    /// Load the user's tiles. Shows a spinner only on the first load; a refresh
    /// keeps the current list visible (pull-to-refresh has its own indicator).
    func load() async {
        if tiles.value == nil {
            tiles = .loading
        }
        do {
            tiles = .loaded(try await repository.fetchTiles())
        } catch {
            tiles = .failed(error)
        }
    }

    /// Create a tile and prepend it to the list. Throws so the form can surface
    /// the failure and stay open.
    func create(title: String, url: String, notes: String?) async throws {
        let tile = try await repository.createTile(title: title, url: url, notes: notes)
        var current = tiles.value ?? []
        current.insert(tile, at: 0)
        tiles = .loaded(current)
    }

    /// Update a tile in place. Throws so the form can surface the failure.
    func update(id: UUID, title: String, url: String, notes: String?) async throws {
        let updated = try await repository.updateTile(id: id, title: title, url: url, notes: notes)
        guard var current = tiles.value,
              let index = current.firstIndex(where: { $0.id == id }) else { return }
        current[index] = updated
        tiles = .loaded(current)
    }

    /// Delete a tile. On failure the tile stays in the list so the user can retry.
    func delete(_ tile: Tile) async {
        do {
            try await repository.deleteTile(id: tile.id)
            guard var current = tiles.value else { return }
            current.removeAll { $0.id == tile.id }
            tiles = .loaded(current)
        } catch {
            // Keep the tile on failure.
        }
    }
}

File: android/app/src/main/java/com/yona/app/core/TileStore.kt
package com.yona.app.core

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Holds the Home tile list as an explicit [LoadState] so the UI renders
 * loading / empty / loaded / error. Mirrors the iOS TileStore.
 */
object TileStore {
    var tiles by mutableStateOf<LoadState<List<Tile>>>(LoadState.Idle)
        private set

    /** Load the list. [showLoading] = false keeps the current grid visible (pull-to-refresh). */
    suspend fun load(showLoading: Boolean = true) {
        if (tiles !is LoadState.Loaded) {
            // Render the cached list instantly on cold launch, then refresh below.
            val cached = TileCache.read()
            tiles = when {
                cached != null -> LoadState.Loaded(cached)
                showLoading -> LoadState.Loading
                else -> tiles
            }
        }
        runCatching { TileRepository.fetchTiles() }.fold(
            onSuccess = {
                tiles = LoadState.Loaded(it)
                TileCache.write(it)
            },
            onFailure = { error ->
                // Don't blow away an existing (or cached) list on a silent refresh failure.
                if (tiles !is LoadState.Loaded) {
                    tiles = LoadState.Failed(error.message ?: "Couldn't load your accounts.")
                }
            },
        )
    }

    /** Reset in-memory state (e.g. on sign-out) so the next user doesn't see stale tiles. */
    fun reset() {
        tiles = LoadState.Idle
    }

    /** Create a tile, then refresh the list so it appears on Home. */
    suspend fun create(draft: TileDraft): Result<Unit> =
        runCatching { TileRepository.createTile(draft) }
            .onSuccess { load(showLoading = false) }
            .map { }

    /** Update a tile, then refresh the list. */
    suspend fun update(id: String, draft: TileDraft): Result<Unit> =
        runCatching { TileRepository.updateTile(id, draft) }
            .onSuccess { load(showLoading = false) }
            .map { }

    /** Delete a tile (sweeping its Storage files first), then refresh the list. */
    suspend fun delete(id: String): Result<Unit> =
        runCatching {
            // Best-effort: the row cascade removes attachment rows, not the files.
            runCatching { AttachmentRepository.deleteTileAttachments(id) }
            TileRepository.deleteTile(id)
        }.onSuccess { load(showLoading = false) }
}
